tests/DataValues/TimeValueCalculatorTest.php

```diff
@@ -209,4 +209,450 @@ public function testGetSecondsForPrecision( $precision, $expected ) {
+						$oldTimestamp,
+						0, 0, 0,
+						$oldPrecision,
+						TimeValue::CALENDAR_GREGORIAN
+					);
+					$timeValue = new TimeValue(
+						$timestamp,
+						0, 0, 0,
+						$precision,
+						TimeValue::CALENDAR_GREGORIAN
+					);
+					yield [ [ $oldTimeValue, $timeValue ] ];
+				}
+				$oldTimestamp = $timestamp;
+				$oldPrecision = $precision;
+			}
+		}
+	}
+
+	/**
+	 * @return \Generator
+	 */
+	public function provideTimestampsAndPrecisions() {
+		foreach ( $this->provideTimestamps() as $timestamp ) {
+			foreach ( $this->provideAllPrecisions() as $precision ) {
+				yield [ [ $timestamp, $precision ] ];
+			}
+		}
+	}
+
+	/**
+	 * @return \Generator of ISO 8601 timestamp arrays [ t1, t2 ] where t1 ≤ t2
+	 */
+	public function provideTimestampPairs() {
+		foreach ( [ '+', '-' ] as $sign ) {
+			$oldTimestamp = null;
+			foreach ( $this->provideTimestampsWithoutSign() as $timestamp ) {
+				$timestamp = $sign . $timestamp;
+				if ( $oldTimestamp !== null ) {
+					yield $sign === '+' ?
+						[ [ $oldTimestamp, $timestamp ] ] :
+						[ [ $timestamp, $oldTimestamp ] ];
+				}
+				$oldTimestamp = $timestamp;
+			}
+		}
+	}
+
+	/**
+	 * @return \Generator of arrays of:
+	 * - a TimeValue,
+	 * - its lower Unix timestamp, and
+	 * - its Unix higher timestamp.
+	 */
+	public function provideTimeValuesAndLowerAndHigherTimestamps() {
+		$timeValuesAndLowerAndHigherTimestamps = [
+			[
+				new TimeValue( // BCE calculations can differ one year from external ones
+					'-0754-11-27T10:49:31Z',
+					0, 0, 0,
+					TimeValue::PRECISION_SECOND,
+					TimeValue::CALENDAR_GREGORIAN
+				),
+				-85901001029,
+				-85901001029
+			],
+			[
+				new TimeValue( // BCE calculations can differ one year from external ones
+					'-0003-01-06T10:49:31Z',
+					0, 0, 0,
+					TimeValue::PRECISION_YEAR,
+					TimeValue::CALENDAR_GREGORIAN
+				),
+				-62230291200,
+				-62198755201
+			],
+			[
+				new TimeValue(
+					'+0002-07-17T02:41:22Z',
+					0, 0, 0,
+					TimeValue::PRECISION_DAY,
+					TimeValue::CALENDAR_GREGORIAN
+				),
+				-62087040000,
+				-62086953601
+			],
+			[
+				new TimeValue(
+					'+2013-03-14T12:51:02Z',
+					0, 0, 0,
+					TimeValue::PRECISION_MONTH,
+					TimeValue::CALENDAR_GREGORIAN
+				),
+				1362096000,
+				1364774399
+			],
+			[
+				new TimeValue(
+					'+7254-11-27T10:49:31Z',
+					0, 0, 0,
+					TimeValue::PRECISION_SECOND,
+					TimeValue::CALENDAR_GREGORIAN
+				),
+				166775539771,
+				166775539771
+			],
+			[
+				new TimeValue(
+					'+190134401-01-15T10:39:59Z',
+					0, 0, 0,
+					TimeValue::PRECISION_DAY,
+					TimeValue::CALENDAR_GREGORIAN
+				),
+				5999999999961600,
+				6000000000047999
+			],
+			[
+				new TimeValue(
+					'+190134401-01-15T10:39:59Z',
+					0, 0, 0,
+					TimeValue::PRECISION_SECOND,
+					TimeValue::CALENDAR_GREGORIAN
+				),
+				5999999999999999,
+				5999999999999999
+			],
+		];
+		yield $timeValuesAndLowerAndHigherTimestamps;
```

Comment:
Instead of yielding the whole array and then iterating over it in the test functions, why not return the array (or yield the individual elements)?